doc/TUTORIAL.md

```diff
@@ -54,177 +46,197 @@ This information allows other humans to help and build on your work.
-  - How to get the keys?
-
-### (4.3) Add Badges to Front Page
-
-Once you are done implementing your badge, you can add it to the collection on [shields.io](http://shields.io/).
-
-First, we make it visible locally at [http://[::1]:8080/][home]. Edit
-[lib/all-badge-examples.js][allbadges] in the right category (Build,
-Downloads, ...) and add your badge:
+Description of the code:
+1. As with the first example, we declare strict mode at the start of each file.
+2. Our badge will query a JSON API so we will extend `BaseJsonService` instead of `BaseService`. This contains some helpers to reduce the need for boilerplate when calling a JSON API.
+3. In this case we are making a version badge, which is a common patten. Instead of directly returning an object in this badge we will use a helper function to format our data consistently. There are a variety of helper functions to help with common tasks in `/lib`.
+4. We perform input validation by defining a schema which we expect the JSON we receive to conform to. This is done using [Joi](https://github.com/hapijs/joi). Defining a schema means we can ensure the JSON we receive meets our expectations and throw an error if we receive unexpected input without having to explicitly code validation checks. The schema also acts as a filter on the JSON object. Any properties we're going to reference need to be validated, otherwise they will be filtered out. In this case our schema declares that we expect to reveive an object which must have a property called 'status', which is a string.
+5. Our module exports a class which extends `BaseJsonService`
+6. As with our previous badge, we need to declare a route. This time we will capture a variable called `gem`.
+7. We can use `defaultBadgeData()` to set a default `color`, `logo` and/or `label`. If `handle()` doesn't return any of these keys, we'll use the default. Instead of explicitly setting the label text when we return a badge object, we'll use `defaultBadgeData()` here to define it declaratively.
+8. Our bage must implement the `async handle()` function. Because our URL pattern captures a variable called `gem`, our function signature is `async handle({ gem })`. We usually seperate the process of generating a badge into 2 stages or concerns: fetch and render. The `fetch()` function is responsible for calling an API endpoint to get data. The `render()` function formats the data for display. In a case where there is a lot of calculation or intermediate steps, this pattern may be thought of as fetch, transform, render and it might be necessary to define some helper functions to assist with the 'transform' step.
+9. The `async fetch()` method is responsible for calling an API endpoint to get data. Extending `BaseJsonService` gives us the helper function `_requestJson()`. Note here that we pass the schema we defined in step 4 as an argument. `_requestJson()` will deal with validating the response against the schema and throwing an error if necessary.
```

Comment:
I know we can also pass headers, query string etc here and I'd like to explain this, but there are so many layers of abstraction between here and [request](https://github.com/request/request) I struggled to work out how to clearly explain how we're wrapping it or what subset of the `request` docs are applicable. Any pointers on how to explain this well?

Reply:
It would be great to reduce the layers of abstraction. Though, meanwhile, we could say these things:

1. `_requestJson` automatically adds an `Accept` header, checks the status code, parses the response as JSON, and returns the parsed response.

2. Options can be passed to `request`, including method, query string, and headers. If headers are provided they will override the ones automatically set by `_requestJson`. There is no need to specify `json`, as the JSON parsing is handled by `_requestJson`. These are all the supported options: https://github.com/request/request#requestoptions-callback

3. Error messages corresponding to each status code can be returned by passing a dictionary of status codes -> messages in `errorMessages`.